lib/improver/utilities/statistical_operations.py

```diff
@@ -0,0 +1,213 @@
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module to contain statistical operations."""
+
+import numpy as np
+import iris
+from iris.exceptions import CoordinateNotFoundError
+
+
+class ProbabilitiesFromPercentiles2D(object):
+    """
+    Generate a 2-dimensional field of probabilties by interpolating a
+    percentiled cube of data to required points.
+
+    Examples:
+
+        Given a reference field of values against a percentile coordinate, an
+        interpolation is performed using another field of values of the same
+        type (e.g. height). This returns the percentile with which these
+        heights would be associated in the reference field. This effectively
+        uses the field of values as a 2-dimensional set of thresholds, and the
+        percentiles looked up correspond to the probabilities of these
+        thresholds being reached.
+
+        Snow-fall level::
+
+            Reference field: Percentiled snow fall level (m ASL)
+            Other field: Orography (m ASL)
+
+            300m ----------------- 30th Percentile snow fall level
+            200m ----_------------ 20th Percentile snow fall level
+            100m ---/-\----------- 10th Percentile snow fall level
+            000m --/---\----------  0th Percentile snow fall level
+            ______/     \_________ Orogaphy
+
+        The orography heights are compared against the heights that
+        correspond with percentile values to find the band in which they
+        fall; this diagram hides the 2-dimensional variability of the snow
+        fall level. The percentile values are then interpolated to the
+        height of the point being considered. This constructs a
+        2-dimensional field of probabilities that snow will be falling at
+        each point in the orography field.
+    """
+
+    def __init__(self, reference_cube, output_name, inverse_ordering=False):
```

Comment:
This plugin feels the wrong way around. The "reference_cube" is actually the latest data which will be updating each day and the cube is the static field of orography. It becomes more obvious further down when we base the new probability field on the "reference_cube".

The ability to set your own output_name is useful but I don't think it should be compulsory to supply this. It should default to "probability_of'+reference_cube.name() if not supplied.